In [1]:
import random
from tqdm import tqdm
import concurrent.futures

In [4]:
#teenie beanie babies simulation
#10 types of babies
#assume equal circulation
#find approximation for how many meals needed to get all 10

def beaniesimulation(Simcounter=5000000):
    results=0
    for i in range(Simcounter):
        beanietypes=set()
        counter=0
        while len(beanietypes)<10:
            counter+=1
            active=random.randint(1,10)
            beanietypes.add(active)
        results+=(counter)
    mean=results/Simcounter
    return mean


In [5]:
beaniesimulation()

29.2828216

Assuming that each meal had an equal chance of a specific teenie beanie, then according to the simulation one will have to buy around 29 meals. The simulation was conducted 5,000,000 times. Each simulation counted how long it took to get all 10 beanies. It then took the average amount and returned 29.282 times. This means on average one needs to buy 29 happy meals to get all 10 beanies. 

In [23]:
"""World series
Consists of 7 games max. First to 4 wins
first 2 at home, next 3 at yankee, and final 2 at home
there is home advantage where 55% chance of winning
how often will the team that starts at home win the series
"""
def Worldseriessim(Simcounter=9000000):
    with tqdm(total=Simcounter, desc="Processing...") as pbar:
        results=0
        for i in range(Simcounter):
            teamwins={'1':0, '2':0}
            for i in range(2):
                active = random.randint(1, 100)
                if active <= 55:
                    teamwins['1'] += 1
                else:
                    teamwins['2'] += 1
            
            for i in range(3):
                active = random.randint(1, 100)
                if active < 45:  
                    teamwins['1'] += 1
                else:
                    teamwins['2'] += 1
            
            for i in range(2):
                active = random.randint(1, 100)
                if active <= 55:
                    teamwins['1'] += 1
                else:
                    teamwins['2'] += 1

            if teamwins['1'] > teamwins['2']:
                results += 1
            
            pbar.update(1)
    mean=results/Simcounter
    return mean
mean=Worldseriessim()*100
print(f'The team which starts at home has a {mean}% chance of winning.')

Processing...: 100%|██████████| 9000000/9000000 [00:40<00:00, 224874.34it/s]

The team which starts at home has a 50.61324444444445% chance of winning.


The theoretical probability of eating a hot dorito for the first round of roulette is 1/6. 

In [27]:
def chiproulette(simcount=9000000):
    with tqdm(total=simcount, desc="Processing...") as pbar:
        results=0
        for i in range(simcount):
            if random.randint(1,6) == 6:
                results+=1
            pbar.update(1)
        final=(results/simcount)*100
    return final, simcount

pain, suffering=chiproulette()
print(f'After {suffering} amount of trials, we have found that on average you have a {pain}% chance of having the painful dorrito.')

Processing...: 100%|██████████| 9000000/9000000 [00:05<00:00, 1676901.86it/s]

After 9000000 amount of trials, we have found that on average you have a 16.660533333333333% chance of having the painful dorrito.


After running a significant amount of trials, the empirical probability should approach the theoretical probability due to the law of large numbers. My empirical probability is 16.6605% which is very close to the 16.6666% theoretical probability.

In [7]:
def roulettegame():
    playercount = int(input('How many players? '))
    players = {'player': [], 'number': []}
    
    print('Loading and spinning...')
    activechamber=random.randint(1,6)
    currentchamber=random.randint(1,6)
    for i in range(playercount):
        playername = input(f'What is the name of player {i + 1}? ')
        players['player'].append(playername)
        players['number'].append(i)

    print('Let’s see who goes first...')
    current_player_index = random.randint(0, playercount - 1) 
    current_player = players['player'][current_player_index]
    while len(players['player']) > 0:
        current_player = players['player'][current_player_index]
        print(f'{current_player} has the gun. Pull the trigger! (Press Enter)')
        input()
        
        if currentchamber==activechamber:
            print(f'BANG! {current_player} has a booboo!')
            players['player'].remove(current_player)
            if current_player_index >= len(players['player']):
                current_player_index = 0
            activechamber=random.randint(1,6)
            currentchamber=random.randint(1,6)

        else:
            print(f'Click! {current_player} survives!')
            current_player_index = (current_player_index + 1) % len(players['player'])            
            currentchamber+=1
            if currentchamber==7:
                currentchamber=1


In [ ]:
roulettegame()

In [52]:
def birthday(people=23, simulations=900000):
    with tqdm(total=simulations, desc="Processing...") as pbar:

        results=0
        currentsim=0
        for i in range(simulations):
            birthdaylist = []
            for i in range(people):
                birthday = random.randint(1, 365)
                if birthday in birthdaylist:
                    results += 1
                    break
                birthdaylist.append(birthday)
            pbar.update(1)
    fin=(results/simulations)*100
    return fin

In [53]:
fin=birthday()
print(fin)

Processing...: 100%|██████████| 900000/900000 [00:33<00:00, 26730.62it/s]

50.800222222222224


In [10]:
# the theoretical probability of someone donating o- or AB- blood is 6.6% + .6% = 7.2%

def bloodtypeprobabilities(simulations=90000000):
    with tqdm(total=simulations, desc="Processing...") as pbar:
        results=0
        for i in range(simulations):
            Type=random.randint(1, 1000)
            if Type <= 66:
                results+=1
            elif Type in [67, 68, 69, 70, 71, 72]:
                results+=1
            pbar.update(1)
    fin=(results/simulations)*100
    return fin
fin = bloodtypeprobabilities()
print(f'{fin}%')

#the empirical probability is very similar to the theoretical probability

Processing...: 100%|██████████| 90000000/90000000 [01:01<00:00, 1455786.14it/s]

7.197706666666667%


In [13]:
#total possibilities 26+26+10+32=94  so the possibilities of a 15 digit password is 94^15=3.95291799 × 10^29. This means you have a 1 in 3.95291799 × 10^29 chance of getting the password first try 
def passwordguessing(simulations=90000000):
    with tqdm(total=simulations, desc="Processing...") as pbar:
        results=0
        for i in range(simulations):
            potential=94**15
            password=random.randint(1, potential)
            passwordguess=random.randint(1, potential)
            if password == passwordguess:
                results+=1
            pbar.update(1)
    fin=(results/simulations)*100
    return fin
fin = passwordguessing()
print(f'{fin}%')

Processing...: 100%|██████████| 90000000/90000000 [03:14<00:00, 463064.66it/s]

0.0%


In [ ]:
import random
import concurrent.futures
from tqdm import tqdm

def simulate_password_guesses(maxsimulations, pbar, idx):
    counts = 0
    password = random.randint(1, maxsimulations)
    guesses = set()
    
    for _ in range(maxsimulations):
        guess = random.randint(1, maxsimulations)
        if guess == password:
            break
        if guess not in guesses:
            counts += 1
            guesses.add(guess)
        
        pbar.update(1)  
    
    return counts

def timetopass(characters=8, sims=1, num_workers=5000):
    maxsimulations = 94 ** characters
    total_counts = 0
    
    with tqdm(total=sims, desc="Simulating...", position=0) as main_pbar:
        with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
            futures = []
            simulation_bars = []
            
            for idx in range(sims):
                sim_bar = tqdm(total=maxsimulations, desc=f"Sim {idx + 1}", position=idx + 1, leave=False)
                simulation_bars.append(sim_bar)
                futures.append(executor.submit(simulate_password_guesses, maxsimulations, sim_bar, idx + 1))
            
            for future in concurrent.futures.as_completed(futures):
                total_counts += future.result()
                main_pbar.update(1)
    
    avg_guesses = total_counts / sims
    return avg_guesses

fin = timetopass()
print(f'It took an average of {fin} guesses to guess the password.')


Sim 1:   0%|          | 4978896/6095689385410816 [00:09<3171531:28:15, 533889.41it/s]


Sim 1:   0%|          | 10021283/6095689385410816 [00:19<3065151:17:58, 552418.75it/s]


Sim 1:   0%|          | 14649884/6095689385410816 [00:29<4074555:42:49, 415566.06it/s]


Sim 1:   0%|          | 19356616/6095689385410816 [00:38<3199989:16:28, 529141.48it/s]


Sim 1:   0%|          | 24064641/6095689385410816 [00:48<3215079:11:46, 526657.96it/s]


Sim 1:   0%|          | 29032433/6095689385410816 [00:58<3460513:44:44, 489305.11it/s]


Sim 1:   0%|          | 33879770/6095689385410816 [01:07<3268608:58:07, 518032.92it/s]


Sim 1:   0%|          | 38817012/6095689385410816 [01:17<3195902:39:43, 529818.09it/s]


Sim 1:   0%|          | 43811126/6095689385410816 [01:27<3027949:49:09, 559205.78it/s]


Sim 1:   0%|          | 48599157/6095689385410816 [01:37<3425253:53:00, 494342.05it/s]


Sim 1:   0%|          | 53538794/6095689385410816 [01:46<3146799:13:28, 538085.50it/s]


Sim 1:   0%|          | 58566040/6095689385410816 [01:56<3106895:49:24, 544996.40it/s]


Sim 1:   0%|          | 63610743/6095689385410816 [02:05<3011619:10:48, 562238.10it/s]


Sim 1:   0%|          | 68295228/6095689385410816 [02:15<3258943:24:56, 519569.33it/s]


Sim 1:   0%|          | 73252747/6095689385410816 [02:24<3121162:39:52, 542505.22it/s]


Sim 1:   0%|          | 78006796/6095689385410816 [02:34<3264488:23:15, 518686.80it/s]


Sim 1:   0%|          | 82681246/6095689385410816 [02:45<3277326:54:20, 516654.91it/s]


Sim 1:   0%|          | 86714599/6095689385410816 [02:56<6087902:47:09, 278133.06it/s]


Sim 1:   0%|          | 91113417/6095689385410816 [03:05<3067529:13:16, 551990.51it/s]


Sim 1:   0%|          | 95865253/6095689385410816 [03:15<3046006:21:01, 555890.84it/s]


Sim 1:   0%|          | 100797572/6095689385410816 [03:25<3709871:05:39, 456416.67it/s]


Sim 1:   0%|          | 105306157/6095689385410816 [03:34<3137490:56:57, 539681.88it/s]


Sim 1:   0%|          | 110179288/6095689385410816 [03:44<3286183:58:41, 515262.39it/s]


Sim 1:   0%|          | 114699283/6095689385410816 [03:53<4465264:32:11, 379204.19it/s]


Sim 1:   0%|          | 119646505/6095689385410816 [04:02<3146821:02:00, 538081.77it/s]


Sim 1:   0%|          | 124496676/6095689385410816 [04:12<3452812:03:56, 490396.52it/s]


Sim 1:   0%|          | 129443122/6095689385410816 [04:21<3790889:07:25, 446662.24it/s]


Sim 1:   0%|          | 134120076/6095689385410816 [04:31<5963470:53:05, 283936.49it/s]


Sim 1:   0%|          | 138633038/6095689385410816 [04:40<3801369:47:28, 445430.75it/s]


Sim 1:   0%|          | 143463532/6095689385410816 [04:50<3094297:42:14, 547215.29it/s]


Sim 1:   0%|          | 148336723/6095689385410816 [04:59<3243936:00:04, 521973.00it/s]


Sim 1:   0%|          | 153265594/6095689385410816 [05:08<3134464:33:26, 540202.95it/s]


Sim 1:   0%|          | 158111255/6095689385410816 [05:18<2990114:58:41, 566281.57it/s]


Sim 1:   0%|          | 162655249/6095689385410816 [05:32<3592163:36:02, 471372.46it/s]


Sim 1:   0%|          | 167624113/6095689385410816 [05:41<3173530:37:08, 533553.07it/s]


Sim 1:   0%|          | 172524751/6095689385410816 [05:51<3032500:30:40, 558366.60it/s]


Sim 1:   0%|          | 177399766/6095689385410816 [06:00<3491718:18:14, 484932.30it/s]


Sim 1:   0%|          | 181811813/6095689385410816 [06:10<3313021:30:43, 511088.44it/s]


Sim 1:   0%|          | 186541587/6095689385410816 [06:19<3206568:40:56, 528055.74it/s]


Sim 1:   0%|          | 191082810/6095689385410816 [06:29<3726219:21:25, 454414.20it/s]


Sim 1:   0%|          | 195516409/6095689385410816 [06:38<4169773:05:07, 406076.53it/s]


Sim 1:   0%|          | 199198730/6095689385410816 [06:48<4006211:57:27, 422655.37it/s]


Sim 1:   0%|          | 203686901/6095689385410816 [06:57<3076434:48:31, 550392.61it/s]


Sim 1:   0%|          | 208389253/6095689385410816 [07:06<3419608:49:33, 495158.10it/s]


Sim 1:   0%|          | 213207715/6095689385410816 [07:16<3013399:34:50, 561905.90it/s]


Sim 1:   0%|          | 218216331/6095689385410816 [07:25<3007064:24:55, 563089.70it/s]


Sim 1:   0%|          | 222982827/6095689385410816 [07:35<3228399:17:53, 524484.99it/s]


Sim 1:   0%|          | 227837301/6095689385410816 [07:44<3102866:07:12, 545704.17it/s]


Sim 1:   0%|          | 232575966/6095689385410816 [07:53<3262774:21:15, 518959.27it/s]


Sim 1:   0%|          | 237474693/6095689385410816 [08:03<3177541:40:55, 532879.55it/s]


Sim 1:   0%|          | 242394378/6095689385410816 [08:12<3254186:21:12, 520328.83it/s]


Sim 1:   0%|          | 247217213/6095689385410816 [08:21<3590326:51:27, 471613.60it/s]


Sim 1:   0%|          | 251951671/6095689385410816 [08:31<3626521:57:29, 466906.58it/s]


Sim 1:   0%|          | 256782549/6095689385410816 [08:40<3066432:40:09, 552187.89it/s]


Sim 1:   0%|          | 261697142/6095689385410816 [08:50<3177596:53:15, 532870.29it/s]


Sim 1:   0%|          | 266495672/6095689385410816 [08:59<3213557:36:41, 526907.30it/s]


Sim 1:   0%|          | 271272534/6095689385410816 [09:08<3287191:02:11, 515104.52it/s]


Sim 1:   0%|          | 276001835/6095689385410816 [09:18<3080010:45:46, 549753.59it/s]


Sim 1:   0%|          | 280762587/6095689385410816 [09:27<3360855:33:32, 503814.27it/s]


Sim 1:   0%|          | 285224731/6095689385410816 [09:37<3496855:57:52, 484219.82it/s]


Sim 1:   0%|          | 290133098/6095689385410816 [09:46<3197588:17:51, 529538.77it/s]


Sim 1:   0%|          | 294148668/6095689385410816 [09:56<5178389:08:42, 326983.34it/s]


Sim 1:   0%|          | 297810972/6095689385410816 [10:06<3647577:13:12, 464211.41it/s]


Sim 1:   0%|          | 302069290/6095689385410816 [10:16<3262257:22:09, 519041.50it/s]


Sim 1:   0%|          | 305880714/6095689385410816 [10:26<5991636:06:58, 282601.77it/s]


Sim 1:   0%|          | 309790677/6095689385410816 [10:35<3975678:16:18, 425901.41it/s]


Sim 1:   0%|          | 314208954/6095689385410816 [10:45<3148729:24:10, 537755.63it/s]


Sim 1:   0%|          | 318879356/6095689385410816 [10:54<3188784:36:00, 531000.73it/s]


Sim 1:   0%|          | 322788480/6095689385410816 [12:17<6111105:42:26, 277077.02it/s]


Sim 1:   0%|          | 327327090/6095689385410816 [12:26<3294489:57:43, 513963.31it/s]


Sim 1:   0%|          | 332024016/6095689385410816 [12:35<3264644:40:14, 518661.95it/s]


Sim 1:   0%|          | 336854682/6095689385410816 [12:44<3100214:54:16, 546170.83it/s]


Sim 1:   0%|          | 341692863/6095689385410816 [12:53<3276992:29:13, 516707.61it/s]


Sim 1:   0%|          | 346581192/6095689385410816 [13:03<3204329:43:14, 528424.69it/s]


Sim 1:   0%|          | 351365238/6095689385410816 [13:12<3159591:26:28, 535906.93it/s]


Sim 1:   0%|          | 356302068/6095689385410816 [13:21<3207686:36:04, 527871.69it/s]


Sim 1:   0%|          | 361158136/6095689385410816 [13:30<3221277:19:06, 525644.58it/s]


Sim 1:   0%|          | 366097123/6095689385410816 [13:39<3208963:16:57, 527661.68it/s]


Sim 1:   0%|          | 370918499/6095689385410816 [13:48<3034318:58:22, 558031.95it/s]


Sim 1:   0%|          | 375783125/6095689385410816 [13:57<3177164:13:11, 532942.85it/s]


Sim 1:   0%|          | 380587127/6095689385410816 [14:07<3432653:43:38, 493276.36it/s]


Sim 1:   0%|          | 385411529/6095689385410816 [14:16<3153051:52:13, 537018.42it/s]


Sim 1:   0%|          | 390322728/6095689385410816 [14:25<3186193:30:05, 531432.55it/s]


Sim 1:   0%|          | 395141147/6095689385410816 [14:34<3187145:55:39, 531273.74it/s]


Sim 1:   0%|          | 400010427/6095689385410816 [14:43<3211868:34:33, 527184.38it/s]


Sim 1:   0%|          | 404857486/6095689385410816 [14:52<3352139:04:16, 505124.31it/s]


Sim 1:   0%|          | 409624934/6095689385410816 [15:01<3204368:41:19, 528418.26it/s]


Sim 1:   0%|          | 414530611/6095689385410816 [15:11<3101841:43:21, 545884.38it/s]


Sim 1:   0%|          | 419279211/6095689385410816 [15:20<3137608:42:19, 539661.60it/s]


Sim 1:   0%|          | 424080144/6095689385410816 [15:29<3185525:09:45, 531544.05it/s]


Sim 1:   0%|          | 428671610/6095689385410816 [15:38<3268740:13:51, 518012.08it/s]


Sim 1:   0%|          | 433395262/6095689385410816 [15:47<3272290:46:22, 517450.02it/s]


Sim 1:   0%|          | 438242605/6095689385410816 [15:56<3077269:43:03, 550243.26it/s]


Sim 1:   0%|          | 443137356/6095689385410816 [16:06<3122260:45:52, 542314.39it/s]


Sim 1:   0%|          | 448017282/6095689385410816 [16:15<3236480:49:40, 523175.33it/s]


Sim 1:   0%|          | 452831495/6095689385410816 [16:24<3176026:50:25, 533133.70it/s]


Sim 1:   0%|          | 457592503/6095689385410816 [16:33<3514051:02:12, 481850.41it/s]


Sim 1:   0%|          | 461987566/6095689385410816 [16:42<3467579:28:55, 488308.04it/s]


Sim 1:   0%|          | 466661049/6095689385410816 [16:52<3653688:02:20, 463435.00it/s]


Sim 1:   0%|          | 471238298/6095689385410816 [17:01<3257950:12:25, 519727.69it/s]


Sim 1:   0%|          | 475707010/6095689385410816 [17:10<3576970:55:23, 473374.53it/s]


Sim 1:   0%|          | 480332637/6095689385410816 [17:19<3320746:43:53, 509899.45it/s]


Sim 1:   0%|          | 484570999/6095689385410816 [17:28<3790583:10:07, 446698.26it/s]


Sim 1:   0%|          | 488836996/6095689385410816 [17:38<3670057:48:33, 461367.91it/s]


Sim 1:   0%|          | 493428738/6095689385410816 [17:47<3265254:30:50, 518565.06it/s]


Sim 1:   0%|          | 498120871/6095689385410816 [17:56<3303010:20:41, 512637.48it/s]


Sim 1:   0%|          | 502894633/6095689385410816 [18:05<3130711:46:09, 540850.46it/s]


Sim 1:   0%|          | 507592081/6095689385410816 [18:14<3329537:42:32, 508553.16it/s]


Sim 1:   0%|          | 512243671/6095689385410816 [18:24<3372882:11:35, 502017.80it/s]


Sim 1:   0%|          | 516919313/6095689385410816 [18:33<3297567:59:32, 513483.55it/s]


Sim 1:   0%|          | 521251786/6095689385410816 [18:42<3426028:00:15, 494230.32it/s]


Sim 1:   0%|          | 525887008/6095689385410816 [18:51<3298828:16:10, 513287.38it/s]


Sim 1:   0%|          | 530666990/6095689385410816 [19:00<3159982:16:26, 535840.63it/s]


Sim 1:   0%|          | 535424281/6095689385410816 [19:10<3163939:50:59, 535170.38it/s]


Sim 1:   0%|          | 540257718/6095689385410816 [19:19<3200954:27:12, 528981.88it/s]


Sim 1:   0%|          | 544895781/6095689385410816 [19:28<3280450:01:45, 516162.99it/s]


Sim 1:   0%|          | 549672492/6095689385410816 [19:37<3421510:01:26, 494882.93it/s]


Sim 1:   0%|          | 552511733/6095689385410816 [19:43<3809384:13:21, 444493.60it/s]

For the sake of pain now it is for password (15 digits) and six digit 2FA

help meeeeee...

there is 94^15 possibilities for the password
the possibilities for the 2FA is 10^6
we have therefore 94^15*10^6 possibilities... not good for a hacker lol